src/gtest/test_checktransaction.cpp

```diff
@@ -391,3 +391,346 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+    mtx.nVersion = 3;
+    mtx.nVersionGroupId = OVERWINTER_VERSION_GROUP_ID;
+    mtx.nExpiryHeight = 0;
+    CTransaction tx(mtx);
+    MockCValidationState state;
+    EXPECT_TRUE(CheckTransactionWithoutProofVerification(tx, state));
+}
+
+TEST(checktransaction_tests, OverwinterExpiryHeight) {
+    CMutableTransaction mtx = GetValidTransaction();
+    mtx.vjoinsplit.resize(0);
+    mtx.fOverwintered = true;
+    mtx.nVersion = 3;
+    mtx.nVersionGroupId = OVERWINTER_VERSION_GROUP_ID;
+    mtx.nExpiryHeight = 0;
+
+    {
+        CTransaction tx(mtx);
+        MockCValidationState state;
+        EXPECT_TRUE(CheckTransactionWithoutProofVerification(tx, state));
+    }
+
+    {
+        mtx.nExpiryHeight = TX_EXPIRY_HEIGHT_THRESHOLD;
+        CTransaction tx(mtx);
+        MockCValidationState state;
+        EXPECT_TRUE(CheckTransactionWithoutProofVerification(tx, state));
+    }
+
+    {
+        mtx.nExpiryHeight = TX_EXPIRY_HEIGHT_THRESHOLD + 1;
+        CTransaction tx(mtx);
+        MockCValidationState state;
+        EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, "bad-tx-expiry-height-too-high", false)).Times(1);
+        CheckTransactionWithoutProofVerification(tx, state);
+    }
+
+    {
+        mtx.nExpiryHeight = std::numeric_limits<uint32_t>::max();
+        CTransaction tx(mtx);
+        MockCValidationState state;
+        EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, "bad-tx-expiry-height-too-high", false)).Times(1);
+        CheckTransactionWithoutProofVerification(tx, state);
+    }
+}
+
+
+// Test that a sprout tx with a negative version number is detected
+// given the new Overwinter logic
+TEST(checktransaction_tests, SproutTxVersionTooLow) {
+    CMutableTransaction mtx = GetValidTransaction();
+    mtx.vjoinsplit.resize(0);
+    mtx.fOverwintered = false;
+    mtx.nVersion = -1;
+
+    CTransaction tx(mtx);
+    MockCValidationState state;
+    EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, "bad-txns-version-too-low", false)).Times(1);
+    CheckTransactionWithoutProofVerification(tx, state);
+}
+
+
+// Test bad Overwinter version numbers in CheckTransactionWithoutProofVerification
+TEST(checktransaction_tests, OverwinterVersionNumber) {
+    CMutableTransaction mtx = GetValidTransaction();
+    mtx.vjoinsplit.resize(0);
+    mtx.fOverwintered = true;
+    mtx.nVersion = 3;
+    mtx.nVersionGroupId = OVERWINTER_VERSION_GROUP_ID;
+    mtx.nExpiryHeight = 0;
+
+    {
+        CTransaction tx(mtx);
+
+        // Evil
+        *const_cast<int32_t*>(&tx.nVersion) = CTransaction::OVERWINTER_MIN_CURRENT_VERSION - 1;
+
+        MockCValidationState state;
+        EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, "bad-tx-overwinter-version-too-low", false)).Times(1);
+        CheckTransactionWithoutProofVerification(tx, state);
+    }
+
+    {
+        CTransaction tx(mtx);
+
+        // Evil
+        *const_cast<int32_t*>(&tx.nVersion) = CTransaction::OVERWINTER_MAX_CURRENT_VERSION + 1;
```

Comment:
> No it's not! It makes the test have undefined behaviour which means, for example, that it could spuriously pass.

Okay, fair enough: for this specific class of evil, it's possible for the test to spuriously pass. But then, is that worse than not having a test at all (which is functionally equivalent to a test that always passes), if the evil is sufficiently documented?

> Also, I disagree in principle that it's necessary to test object states (for constant objects) that the constructor rejects.

The only reason the constructor rejects it is because the constructor calls `UpdateHash()` which performs a serialization. So it's not quite as direct as that, but I can see the reasoning.

Reply:
Yes, it is worse. I don't want to have to think about undefined behaviour in the test suite. It complicates reasoning about what test results tell you.

Reply:
In general we should have zero tolerance for undefined behaviour. We've been too lax about this :-/